### PR TITLE
Ports: Update backward-cpp

### LINKED ISSUE #19708: Ports/backward-cpp "sometimes" not building
I noticed some weird behaviour when trying to build backward-cpp port. On a clean serenity build...
```
$ Meta/serenity.sh delete
$ Meta/serenity.sh build
$ pushd Ports/backward-cpp; ./package.sh; popd
```

the build of backward-cpp fails with error messages "undefined reference to `sframe_...' "
```
=> Building backward-cpp...
[backward-cpp/build] + make -j24
[backward-cpp/build] [ 21%] Built target backward_object
[backward-cpp/build] [ 28%] Built target backward
[backward-cpp/build] [ 28%] Built target test_main
[backward-cpp/build] [ 42%] Linking CXX executable test_rectrace
[backward-cpp/build] [ 50%] Linking CXX executable test_stacktrace
[backward-cpp/build] [ 57%] Linking CXX executable test_select_signals
[backward-cpp/build] [ 64%] Built target test_test
[backward-cpp/build] [ 71%] Linking CXX executable test_suicide
[backward-cpp/build] /home/moormaster/tmp/git/serenity/Toolchain/Local/x86_64/lib/gcc/x86_64-pc-serenity/13.1.0/../../../../x86_64-pc-serenity/bin/ld: /home/moormaster/tmp/git/serenity/Toolchain/Local/x86_64/lib/gcc/x86_64-pc-serenity/13.1.0/../../../../x86_64-pc-serenity/bin/ld: /home/moormaster/tmp/git/serenity/Toolchain/Local/x86_64/lib/gcc/x86_64-pc-serenity/13.1.0/../../../../x86_64-pc-serenity/bin/ld: /home/moormaster/tmp/git/serenity/Toolchain/Local/x86_64/lib/gcc/x86_64-pc-serenity/13.1.0/../../../../x86_64-pc-serenity/bin/ld: /home/moormaster/tmp/git/serenity/Build/x86_64/Root/usr/local/lib/libbfd.a(elfxx-x86.o): in function `/home/moormaster/tmp/git/serenity/Build/x86_64/Root/usr/local/lib/libbfd.a(elfxx-x86.o)_bfd_x86_elf_write_sframe_plt: in function `':
[backward-cpp/build] /home/moormaster/tmp/git/serenity/Build/x86_64/Ports/binutils/binutils-2.40/bfd/elfxx-x86.c:1975: undefined reference to `sframe_encoder_write_bfd_x86_elf_write_sframe_plt':
[backward-cpp/build] '
[backward-cpp/build] /home/moormaster/tmp/git/serenity/Build/x86_64/Ports/binutils/binutils-2.40/bfd/elfxx-x86.c:1975: undefined reference to `/home/moormaster/tmp/git/serenity/Toolchain/Local/x86_64/lib/gcc/x86_64-pc-serenity/13.1.0/../../../../x86_64-pc-serenity/bin/ldsframe_encoder_write: '
[backward-cpp/build] /home/moormaster/tmp/git/serenity/Toolchain/Local/x86_64/lib/gcc/x86_64-pc-serenity/13.1.0/../../../../x86_64-pc-serenity/bin/ld/home/moormaster/tmp/git/serenity/Build/x86_64/Ports/binutils/binutils-2.40/bfd/elfxx-x86.c:: 1981: undefined reference to `sframe_encoder_free'
...
```

But if I build the gdb Port first
```
$ Meta/serenity.sh delete
$ Meta/serenity.sh build
$ pushd Ports/gdb; ./package.sh; popd
$ pushd Ports/backward-cpp; ./package.sh; popd
```
the build of backward-cpp suddenly succeeds. 

Is it missing a dependency in its package.sh? If yes - is gdb itself the missing dependency or is it just a side effect that building gdb helps to make the build of backward-cpp succeed?

### PR DESCRIPTION
**Pull request by LucasChollet**
As mentioned in #19720, the problem with `backward-cpp` comes from upstream. This issue is now fixed, so we just need to update our version.

Fixes #19708 and supersedes #19720

### CODE REVIEW

File: Ports/backward-cpp/patches/0002-backward-Pretend-to-be-Linux-with-some-modifications.patch [hunk 1]
```diff
@@ -58,12 +58,13 @@ index ac7ad51738f8d13f07a6162d421b191696c9f84a..03f4f7546c4e59f3b17529c61a0f387e
...
-@@ -4245,7 +4251,7 @@ public:
-     printer.address = true;
+@@ -4277,7 +4283,8 @@ public:
      printer.print(st, stderr);
  
--#if _XOPEN_SOURCE >= 700 || _POSIX_C_SOURCE >= 200809L
-+#if (_XOPEN_SOURCE >= 700 || _POSIX_C_SOURCE >= 200809L) && !defined(__serenity__)
+ #if (defined(_XOPEN_SOURCE) && _XOPEN_SOURCE >= 700) || \
+-    (defined(_POSIX_C_SOURCE) && _POSIX_C_SOURCE >= 200809L)
++    (defined(_POSIX_C_SOURCE) && _POSIX_C_SOURCE >= 200809L) &&
```
Comment (moormaster) [inline Ports/backward-cpp/patches/0002-backward-Pretend-to-be-Linux-with-some-modifications.patch:66]: This doesn't build for me. Did you try to do a ./package.sh clean && ./package.sh ?

```
[backward-cpp/build] [ 21%] Building CXX object CMakeFiles/test_main.dir/test/_test_main.cpp.o
[backward-cpp/build] In file included from .../git/serenity/Build/x86_64/Ports/backward-cpp/bombela-backward-cpp-65a769f/backward.cpp:36:
[backward-cpp/build] .../git/serenity/Build/x86_64/Ports/backward-cpp/bombela-backward-cpp-65a769f/backward.hpp:4286:64: error: operator '&&' has no right operand
[backward-cpp/build]  4286 |     (defined(_POSIX_C_SOURCE) && _POSIX_C_SOURCE >= 200809L) &&
[backward-cpp/build]       |                                                                ^
[backward-cpp/build] In file included from .../git/serenity/Build/x86_64/Ports/backward-cpp/bombela-backward-cpp-65a769f/backward.cpp:36:
[backward-cpp/build] .../git/serenity/Build/x86_64/Ports/backward-cpp/bombela-backward-cpp-65a769f/backward.hpp:4286:64: error: operator '&&' has no right operand
[backward-cpp/build]  4286 |     (defined(_POSIX_C_SOURCE) && _POSIX_C_SOURCE >= 200809L) &&
[backward-cpp/build]       |                                                                ^
```

Comment (LucasChollet) [reply Ports/backward-cpp/patches/0002-backward-Pretend-to-be-Linux-with-some-modifications.patch:66]: Oh, I messed up a rebase or something, I was supposed to have already fixed it.
I pushed a patch but didn't test it yet, I'm in the middle of a rebuild

Comment (LucasChollet) [reply Ports/backward-cpp/patches/0002-backward-Pretend-to-be-Linux-with-some-modifications.patch]: Looks like it did the trick!

**Review by moormaster — COMMENTED** [review]
It builds fine now in combination with and without port gdb being installed.
But it still seems to have an issue with zstd. If the zstd port is installed the build still fails

```
$ Meta/serenity.sh delete
$ Meta/serenity.sh build
$ pushd Ports/zstd/; ./package.sh; popd
$ pushd Ports/backward-cpp/; ./package.sh; popd
```

```
[backward-cpp/build] [ 92%] Linking CXX executable test_select_signals
[backward-cpp/build] .../git/serenity/Toolchain/Local/x86_64/lib/gcc/x86_64-pc-serenity/13.1.0/../../../../x86_64-pc-serenity/bin/ld: .../git/serenity/Build/x86_64/Root/usr/local/lib/libbfd.a(compress.o): in function `decompress_contents':
[backward-cpp/build] .../git/serenity/Build/x86_64/Ports/binutils/binutils-2.40/bfd/compress.c:517: undefined reference to `ZSTD_decompress'
[backward-cpp/build] .../git/serenity/Toolchain/Local/x86_64/lib/gcc/x86_64-pc-serenity/13.1.0/../../../../x86_64-pc-serenity/bin/ld: .../git/serenity/Build/x86_64/Ports/binutils/binutils-2.40/bfd/compress.c:519: undefined reference to `ZSTD_isError'
[backward-cpp/build] .../git/serenity/Toolchain/Local/x86_64/lib/gcc/x86_64-pc-serenity/13.1.0/../../../../x86_64-pc-serenity/bin/ld: .../git/serenity/Build/x86_64/Root/usr/local/lib/libbfd.a(compress.o): in function `bfd_compress_section_contents':
[backward-cpp/build] .../git/serenity/Build/x86_64/Ports/binutils/binutils-2.40/bfd/compress.c:656: undefined reference to `ZSTD_compress'
[backward-cpp/build] .../git/serenity/Toolchain/Local/x86_64/lib/gcc/x86_64-pc-serenity/13.1.0/../../../../x86_64-pc-serenity/bin/ld: .../git/serenity/Build/x86_64/Ports/binutils/binutils-2.40/bfd/compress.c:661: undefined reference to `ZSTD_isError'
[backward-cpp/build] .../git/serenity/Toolchain/Local/x86_64/lib/gcc/x86_64-pc-serenity/13.1.0/../../../../x86_64-pc-serenity/bin/ld: .../git/serenity/Build/x86_64/Root/usr/local/lib/libbfd.a(compress.o): in function `decompress_contents':
[backward-cpp/build] .../git/serenity/Build/x86_64/Ports/binutils/binutils-2.40/bfd/compress.c:517: undefined reference to `ZSTD_decompress'
[backward-cpp/build] .../git/serenity/Toolchain/Local/x86_64/lib/gcc/x86_64-pc-serenity/13.1.0/../../../../x86_64-pc-serenity/bin/ld: .../git/serenity/Build/x86_64/Ports/binutils/binutils-2.40/bfd/compress.c:519: undefined reference to `ZSTD_isError'
[backward-cpp/build] .../git/serenity/Toolchain/Local/x86_64/lib/gcc/x86_64-pc-serenity/13.1.0/../../../../x86_64-pc-serenity/bin/ld: .../git/serenity/Build/x86_64/Root/usr/local/lib/libbfd.a(compress.o): in function `bfd_compress_section_contents':
[backward-cpp/build] .../git/serenity/Build/x86_64/Ports/binutils/binutils-2.40/bfd/compress.c:656: undefined reference to `ZSTD_compress'
[backward-cpp/build] .../git/serenity/Toolchain/Local/x86_64/lib/gcc/x86_64-pc-serenity/13.1.0/../../../../x86_64-pc-serenity/bin/ld: .../git/serenity/Build/x86_64/Ports/binutils/binutils-2.40/bfd/compress.c:661: undefined reference to `ZSTD_isError'
[backward-cpp/build] collect2: error: ld returned 1 exit status
```

**Comment by LucasChollet** [comment]
I start to understand why all distribution ship a `.so` for `libbfd`, maybe we should do the same.

Applying this patch should work for Serenity, but I'm really not a fan and that's not upstreamable (having libzstd around doesn't mean that binutils has been build with support for it):
```cmake
if (LIBZSTD_LIBRARY)
	list(APPEND _BACKWARD_LIBRARIES ${LIBZSTD_LIBRARY})
endif()
```

Anyway, it looks unrelated to the original issue (I pretty sure that it never worked), and I don't know what we should do. So I think that we can merge that and feel free to add a patch for that problem.

**Review by moormaster — APPROVED** [review]
Seems good to me regarding the scope of the PR

**Review by linusg — APPROVED** [review]

**Pull request merged**